File: JavaScript/1286-Iterator for Combination.js
/**
 * @param {string} characters
 * @param {number} combinationLength
 */
var CombinationIterator = function(characters, combinationLength) {
    this.N = characters.length;
    this.bitmask = (1<<this.N)-1;
    this.k = combinationLength;
    this.characters = characters;
};

/**
 * @return {string}
 */
CombinationIterator.prototype.next = function() {
    this.hasNext();
    let res = new Array();
    for (let i=0; i<this.N; i++){
        if ( (this.bitmask >>(this.N-1-i)) & 1 ){
            res.push(this.characters[i]);
        }
    }
    this.bitmask--;
    return res.join("");
};

/**
 * @return {boolean}
 */
CombinationIterator.prototype.hasNext = function() {
    while (this.bitmask > 0 && this.bitmask.toString(2).split('1').length - 1 !== this.k){
        this.bitmask--;
    }
    return this.bitmask !== 0;
};

/** 
 * Your CombinationIterator object will be instantiated and called as such:
 * var obj = new CombinationIterator(characters, combinationLength)
 * var param_1 = obj.next()
 * var param_2 = obj.hasNext()
 */

File: 1286-Iterator for Combination.py
#O(N*2^N/C(N,k) ) for each query
class CombinationIterator(object):

    def __init__(self, characters, combinationLength):
        """
        :type characters: str
        :type combinationLength: int
        """
        self.N = len(characters)
        self.bitmask = (1<<self.N)-1
        self.k = combinationLength
        self.characters = characters
        
    def next(self):
        """
        :rtype: str
        """
        self.hasNext()
        res = []

        for i in range(self.N):
            if (self.bitmask >> (self.N-1-i)) & 1:
                res.append(self.characters[i])
        self.bitmask -= 1
        
        return "".join(res)

    def hasNext(self):
        """
        :rtype: bool
        """
        while self.bitmask and bin(self.bitmask).count("1") != self.k:
            self.bitmask-=1
        return self.bitmask != 0